solutions/ceoi-04-TwoSawmills.mdx

```diff
@@ -0,0 +1,141 @@
+---
+id: ceoi-04-TwoSawmills
+source: CEOI 04
+title: Two Sawmills
+author: Dustin Miao
+---
+
+# Explaination
+
+Let us define a function $f(a,b)$ which denotes the cost when we place sawmills at $a$ and $b$, $a < b$. Let $T$ be the cost of transporting all trees to the base of the hill (i.e. cost without building any sawmills). Denote two arrays $W$ and $D$ that represent the prefix sum of the weight and distance, respectively. 
+
+$$f(a, b) = T - W[a] \cdot (D[b] - D[a]) - W[b] \cdot (D[n + 1] - D[b])$$
+
+If we hold $b$ as constant, we can rearrange the equation into:
+
+$$
+f(a, b) = T - W[a] \cdot D[b] + W[a] \cdot D[a] - W[b] \cdot D[n + 1] + W[b] \cdot D[b]
+$$
+
+$$
+f(a, b) = \{(T + W[a] \cdot D[a]) - W[a] \cdot D[b]\} - W[b] \cdot D[n +1] + W[b] \cdot D[b]
+$$
+
+Thus, $f$ can be decomposed into four parts: $y_a = T + W[a] \cdot D[a]$, $m_a = -W[a]$, $x = D[b]$, $c = -W[b] \cdot D[n + 1] + W[b] \cdot D[b]$. Thus, 
+
+$$
+f(a, b) = (y_a + m_ax) + c
+$$
+
+Define another function $g(b)$ that denotes the minimum cost if we place the second sawmill at $b$. This can simply be expressed in terms of $f(a,b)$.
+
+$$
+g(b) = \min_{1 \leq a < b} (y_a + m_a x) + c
+$$
+
+Because $x$ and $c$ are functions of $b$, we can hold them constant. Note that $y_a + m_a x$ forms a linear function of $x$. We can use [convex hull trick](https://usaco.guide/plat/convex-hull-trick?lang=cpp) to efficiently query the minimum value of a group of linear functions. The answer is simply $\max_{2 \leq b \leq N}g(b)$. 
```

Comment:
> We can use convex hull trick

bro you sound like an indian youtuber

Reply:
how...?

Reply:
```suggestion
Because $x$ and $c$ are functions of $b$, we can hold them constant. Note that $y_a + m_a x$ forms a linear function of $x$. We can use the [convex hull trick](https://usaco.guide/plat/convex-hull-trick?lang=cpp) to efficiently query the minimum value of a group of linear functions. The answer is simply $\max_{2 \leq b \leq N}g(b)$. 
```